### PR TITLE
Merge/ecouple launcher aten 2 release 3.2.2

### PR DESCRIPTION
**Pull request by zhanwei33**
# New contributor declaration
- [ ] I am not making a trivial change, such as fixing a typo in a comment.

- [ ] I have written a PR description following these
  [rules](https://cbea.ms/git-commit/#why-not-how).

- [ ] I have run `pre-commit run --from-ref origin/main --to-ref HEAD`.

- Select one of the following.
  - [ ] I have added tests.
    - `/test` for `lit` tests
    - `/unittest` for C++ tests
    - `/python/test` for end-to-end tests
  - [ ] This PR does not need a test because `FILL THIS IN`.

- Select one of the following.
  - [ ] I have not added any `lit` tests.
  - [ ] The `lit` tests I have added follow these [best practices](https://mlir.llvm.org/getting_started/TestingGuide/#filecheck-best-practices),
    including the "tests should be minimal" section. (Usually running Python code
    and using the instructions it generates is not minimal.)

# compilation pipeline optimization.

- **Compilation Pipeline Optimization**: Extracted `torch_npu`-related workspace/sync lock/async launch capabilities from the launcher into a standalone `npu_utils.so`, transforming the launcher into a lightweight bridge layer that reduces compile-time heavy dependencies.

### Overview

| Scenario         | Metric                                                 | Old Architecture | New Architecture | Absolute Change | Relative Change |
| ---------------- | ------------------------------------------------------ | ---------------- | ---------------- | --------------- | --------------- |
| First Compile    | `total_compile + npu_utils_build_so + launcher_build`  | 16.266s          | 9.121s           | -7.145s         | -43.93%         |
| Second Compile   | `total_compile + launcher_build`                       | 3.503s           | 2.621s           | -0.882s         | -25.19%         |

### First Compile `softmax` Phase Breakdown

| Phase                      | Old Arch(s) | New Arch(s) | Change(s)  | Notes                                        |
| -------------------------- | ----------: | ----------: | ---------: | -------------------------------------------- |
| `total_compile`            |       2.415 |       2.406 |     -0.009 | Triton kernel core compilation nearly unchanged  |
| `npu_utils_build_so`       |       0.809 |       6.168 |     +5.359 | `torch_npu` dependency cost centralized in `npu_utils.so` build |
| `launcher_build`           |      13.042 |       0.548 |    -12.495 | Launcher build cost reduced significantly    |

Launcher detail:

| Sub-phase                    | Old Arch(s) | New Arch(s) | Change(s)   |
| ---------------------------- | ----------: | ----------: | ---------: |
| `launcher_precompile_header` |      11.615 |       0.000 |    -11.615 |
| `launcher_build_stub`        |      13.040 |       0.546 |    -12.494 |
| `launcher_load_module`       |       0.002 |       0.001 |     -0.001 |

### Second Compile `test_sum` Phase Breakdown

| Phase             | Old Arch(s) | New Arch(s) | Change(s) | Notes                                |
| ----------------- | ----------: | ----------: | --------: | ------------------------------------ |
| `total_compile`   |       2.048 |       2.072 |    +0.024 | Triton kernel core compilation nearly unchanged   |
| `launcher_build`  |       1.456 |       0.549 |    -0.907 | Launcher build continues to benefit  |

Launcher detail:

| Sub-phase                    | Old Arch(s) | New Arch(s) | Change(s) |
| ---------------------------- | ----------: | ----------: | --------: |
| `launcher_precompile_header` |       0.000 |       0.000 |     0.000 |
| `launcher_build_stub`        |       1.453 |       0.548 |    -0.905 |
| `launcher_load_module`       |       0.002 |       0.001 |    -0.001 |

On second compile, `npu_utils.so` has already been cached, so there is no additional `npu_utils_build_so` overhead.

### Performance Analysis

- First compile gain driver: removed `launcher_precompile_header` (11.615s → 0s), launcher build dropped from 13.04s to 0.55s
- Second compile gain driver: launcher build consistently shortened to ~0.55s (1.453s → 0.548s)
- Cost: `npu_utils_build_so` first build increased from 0.809s to 6.168s, but the launcher-side gain is larger, resulting in significant overall compile time improvement
- `Kernel compile` core (`stage_npubin`) is essentially unchanged, confirming the optimization did not impact the core compilation flow

### npu\_utils.so Extraction

Centralized `torch_npu`-related capabilities into `npu_utils.cpp`:

1. **Added `USE_TORCH_NPU` conditional compilation block**: centralized management of `ATen/ATen.h`, `torch_npu/csrc/core/npu/NPUWorkspaceAllocator.h`, `torch_npu/csrc/framework/OpCommand.h`, and other dependencies
2. **Added three `extern "C"` exported interfaces**:
   - `triton_allocate_workspace`: workspace allocation
   - `triton_allocate_sync_block_lock`: sync block lock allocation
   - `triton_async_launch`: taskqueue async launch
3. **Added tensor lifetime retention mechanism (opaque handle + RAII guard)**: prevents raw pointers returned from being invalidated during launch, while also solving the ever-growing global `retained_tensors` vector memory leak risk

### Launcher Lightweighting

The launcher no longer directly depends on `torch_npu`, instead loading `npu_utils.so` at runtime via `dlopen/dlsym`:

- No longer includes `precompiled.h`, replaced with a lightweight header
- No longer directly constructs `at_npu::native::OpCommand`, instead forwards through `npu_utils.so`
- No longer carries `torch_npu` heavy dependencies at compile time

#### Background

After decoupling the launcher, the underlying `at::Tensor` objects for workspace and sync block lock changed from temporary launcher-side objects to internally allocated objects within `npu_utils.so`. The previous implementation used a global `retained_tensors` vector inside `npu_utils.cpp` to hold all allocated tensors. While this approach avoided dangling pointers, the vector only grew without shrinking, causing continuous accumulation of unused workspace/sync lock tensors during long training runs or frequent compilation.

#### Solution

Replace the process-level global `retained_tensors` container with per-allocation releasable opaque handles, with the launcher side using `std::shared_ptr` custom deleters to achieve RAII automatic release. RAII (Resource Acquisition Is Initialization) binds resource lifetime to object lifetime: `shared_ptr` holds the handle upon construction (resource acquisition), and when leaving scope, it automatically destructs and invokes the deleter to release the underlying tensor, guaranteeing no resource leaks regardless of normal return or exception exit.

**New struct in `npu_utils.cpp`:**

```cpp
struct RetainedTensorHandle {
  explicit RetainedTensorHandle(at::Tensor tensor) : tensor(std::move(tensor)) {}
  at::Tensor tensor;
};
```

**Allocation interfaces changed from returning only a device pointer to also returning a handle:**

```cpp
extern "C" void* triton_allocate_workspace(uint64_t size, void **handle);
extern "C" void* triton_allocate_sync_block_lock(uint64_t size, void* stream, void **handle);
extern "C" void triton_release_retained_tensor(void *handle);
```

The `handle` points to a `RetainedTensorHandle`, which holds the actual `at::Tensor`. The launcher still passes the returned `void*` device pointer to runtime args, but also saves the handle.

**New release function pointer in launcher generated code:**

```cpp
typedef void (*triton_release_retained_tensor_t)(void*);
static triton_release_retained_tensor_t g_release_retained_tensor = nullptr;
```

**RAII guard constructed immediately after workspace allocation:**

```cpp
void *workspace_handle = NULL;
workspace_addr_ptr = g_allocate_workspace(totalWorkSpaceSize, &workspace_handle);
std::shared_ptr<void> workspace_handle_guard(workspace_handle, release_npu_tensor_handle);
```

**Sync block lock allocated inside `launch_call`, using the same guard pattern:**

```cpp
void *syncBlockLock_handle = NULL;
syncBlockLock_ptr = g_allocate_sync_block_lock(syncBlockLockSize, stream, &syncBlockLock_handle);
std::shared_ptr<void> syncBlockLock_handle_guard(syncBlockLock_handle, release_npu_tensor_handle);
```

When `TRITON_ENABLE_TASKQUEUE=true`, the `launch_call` lambda captures by `[=]`, so `workspace_handle_guard` stays alive along with the asynchronously submitted callable; in the non-taskqueue path, `_launch` synchronously waits for `rtStreamSynchronize(stream)` before returning, after which the guard destructs and releases the handle.

#### Lifecycle

**workspace:**
1. Launcher calls `g_allocate_workspace(size, &workspace_handle)`
2. `npu_utils.so` creates a `RetainedTensorHandle`, placing the tensor inside
3. Launcher obtains the device pointer `workspace_addr_ptr` and creates `workspace_handle_guard`
4. Kernel launch uses `workspace_addr_ptr`
5. After launch completes, guard destructs, calling `triton_release_retained_tensor(handle)`
6. Handle is deleted, `at::Tensor` destructs, releasing underlying device memory

**sync block lock:**
1. Launcher calls `g_allocate_sync_block_lock(size, stream, &syncBlockLock_handle)` inside the launch callable
2. `npu_utils.so` creates a handle and returns the device pointer
3. Launcher initializes lock data using `rtMemcpy`
4. Kernel launch uses `syncBlockLock_ptr`
5. After callable completes, guard destructs and releases the handle

#### Error Handling

- Allocation interfaces return `nullptr` when receiving a null handle pointer
- `*handle` is set to `nullptr` before allocation to prevent callers from erroneously freeing an uninitialized value
- On workspace allocation failure, prints `Error: workspace allocation failed` and returns
- On sync block lock allocation failure, follows the existing taskqueue/non-taskqueue return strategy
- `release_npu_tensor_handle` prints an error and returns when the release function pointer does not exist, preventing null function pointer invocation

#### Files Involved

- `third_party/ascend/backend/npu_utils.cpp` — removed global `retained_tensors` and mutex, added `RetainedTensorHandle`, allocation interfaces return both device pointer and opaque handle, added `triton_release_retained_tensor`
- `third_party/ascend/backend/driver.py` — launcher generated code gains release function pointer and `release_npu_tensor_handle`, workspace/sync block lock uses `std::shared_ptr<void>` guard to manage handles, `NPUUtils.load_binary` compatible with both old and new call signatures
- `third_party/ascend/backend/backend_register.py` — torch_npu backend workspace/sync block lock allocation template passes handle address
- `third_party/ascend/unittest/pytest_ut/test_backend_npu_utils_lifetime.py` — covers global vector removal, release interface existence, launcher guard generation, and `load_binary` compatibility

### CODE REVIEW

**Comment by zhanwei33** [comment]
Superseded by #214.
